src/session_pool/session_pool.js

```diff
@@ -0,0 +1,148 @@
+     */
+    constructor(options) {
+        super();
+
+        const opts = _.defaults({}, options, SESSION_POOL_DEFAULTS);
+
+        // Pool Configuration
+        this.maxPoolSize = opts.maxPoolSize;
+        this.createSessionFunction = opts.createSessionFunction || this._createSessionFunction;
+
+        // Session configuration
+        // @TODO: Maybe options.sessionOptions / this.sessionOptions?
+        this.maxSessionAgeSecs = opts.maxSessionAgeSecs;
+        this.maxSessionReuseCount = opts.maxSessionReuseCount;
+
+        // Session storage
+        this.persistStateKeyValueStoreId = opts.persistStateKeyValueStoreId;
+        this.persistStateKey = opts.persistStateKey;
+
+        // Statistics
+        this.activeSessions = 0;
+        this.blockedSessions = 0;
+
+        // Operative states
+        this.storage = null;
+        this.sessions = [];
+
+        // Maybe we can add onSessionRetired function to configuration ?
+    }
+
+
+    async initialize() {
+        this.storage = await openKeyValueStore(this.persistStateKeyValueStoreId);
+
+        // Load sessions from storage
+        const loadedSessions = this.storage.getValue(this.persistStateKey) || [];
+
+        // Invalidate old sessions and load active sessions only
+        for (const [sessionName, sessionObject] of Object.entries(loadedSessions)) {
+            const session = Session.recreateSession(sessionObject);
+            const isValid = !session.isExpired() || !session.isBlocked() || session.usedCount >= session.maxReuseCount;
+
+            if (isValid) {
+                this.sessions[sessionName] = session;
+            }
+        }
+    }
+
+    async retrieveSession() {
+        // If we have enough space for session. Return newly created session.
+        if (this.sessions.length < this.maxPoolSize) {
+            const newSession = await this.createSessionFunction(this);
+            this._addSession(newSession);
+            return newSession;
+        }
+
+
+        // For example that developer can plug different picking algorithms such as the Lukášův
+        // Maybe a should have pickSession function to be customizable.
+        const pickedSession = this.sessions[this._getRandomIndex()];
+
+        // If session can be used return the session
+        if (pickedSession.isUsable()) {
+            return pickedSession;
+        }
+
+        //  otherwise remove old session and return newly created session
+        this._removeSession(pickedSession);
+
+        const newSession = this.createSessionFunction(this);
+        this._addSession(newSession);
+        return newSession;
+    }
+
+    getStats() {
+        return {
+            activeSessions: this.activeSessions,
+            blockedSessions: this.blockedSessions,
+        };
+    }
+
+    getState() {
+        return {
+            ...this.getStats(),
+            sessions: this.sessions.map(session => session.getState()),
+        };
+    }
+
+    async persistState() {
```

Comment:
When is this called? E.g. have a look at `RequestList`, the peristState function is called as `events.on(ACTOR_EVENT_NAMES_EX.PERSIST_STATE, this.persistState.bind(this));`

Reply:
Plus you should probably store `isStatePersisted` variable, to avoid unnecessary calls to `setValue()`, same way as RequestList does it

Reply:
IMHO the SessionPool is changing a lot so there is a high chance that it'll change with next persist state call. The problem is not an easy way to say that the SessionPool has changed.

Reply:
Alright, we don't need to track whether the state was change or not, but still, we should make sure that state is persisted on the global ACTOR_EVENT_NAMES_EX.PERSIST_STATE event. Same way as it is for RequestList

Reply:
Yeah, I forgot to do it in the first commit :(. It should be done in the newer one.